var/spack/repos/builtin/packages/upcxx/package.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+import llnl.util.tty as tty
+import os
+
+
+class Upcxx(Package):
+    """UPC++ is a C++ library that supports Partitioned Global Address Space
+    (PGAS) programming, and is designed to interoperate smoothly and
+    efficiently with MPI, OpenMP, CUDA and AMTs. It leverages GASNet-EX to
+    deliver low-overhead, fine-grained communication, including Remote Memory
+    Access (RMA) and Remote Procedure Call (RPC)."""
+
+    homepage = "https://upcxx.lbl.gov"
+
+    version('2019.3.2', '844722cb0e8c0bc649017fce86469457')
+
+    variant('cuda', default=False, 
+            description='Builds a CUDA-enabled version of UPC++')
+    depends_on('cuda', when='+cuda')
+
+    def url_for_version(self, version):
+        if version > Version('2019.3.2'):
+            url = "https://bitbucket.org/berkeleylab/upcxx/downloads/upcxx-{0}.tar.gz"
+        else:
+            url = "https://bitbucket.org/berkeleylab/upcxx/downloads/upcxx-{0}-offline.tar.gz"
+        return url.format(version)
+
+    def setup_environment(self, spack_env, run_env):
+        if 'platform=cray' in self.spec:
+            if (os.environ.get('CROSS') is None):
+                spack_env.set('CROSS', 'cray-aries-slurm')
+                tty.msg("Detected a Cray system, "
+                        "assuming CROSS=cray-aries-slurm")
+                spack_env.set('GASNET_CONFIGURE_ARGS', '--enable-mpi=probe')
+                spack_env.set('NOBS_DEBUG', 'True')
+        if '+cuda' in self.spec:
+            spack_env.set('UPCXX_CUDA', '1')
+            spack_env.set('UPCXX_CUDA_NVCC',
+                          join_path(self.spec['cuda'].prefix.bin, 'nvcc'))
+
+    def setup_dependent_package(self, module, dep_spec):
+        dep_spec.upcxx = join_path(self.prefix.bin, 'upcxx')
+
+    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+        spack_env.set('UPCXX_INSTALL', self.prefix)
+        spack_env.set('UPCXX', join_path(self.prefix.bin, 'upcxx'))
+        if 'platform=cray' in self.spec:
+            spack_env.set('UPCXX_GASNET_CONDUIT', 'aries')
+
+    def install(self, spec, prefix):
+        installsh = which('install', path='.')
+        installsh(prefix)
```

Comment:
I would use:
```python
install_tree('.', prefix)
```

Reply:
I am not sure I understand how `install_tree` should be used. Are you suggesting to rename `installsh` to `install_tree`, or to use `install_tree` from `llnl.util.filesystem`? 

UPC++ comes with its own `install` script, which is captured by `which` in the `installsh` variable, and is then called.

Reply:
Ah, I didn't realize it comes with it's own `install` script, since `install` is also a UNIX command.

In that case, I would use:
```python
installsh = Executable('install')
```
No need to use `which` if you know exactly where the script is located.